e2e/engine/cmd.go

```diff
@@ -285,8 +253,62 @@ func (e *engineCmd) Run(ctx context.Context, eventC chan *node.Event) error {
 				}
 			}
 		}
+		if e.conf.CheckAppHashAfterCmd {
+			if err := checkapphash(e.conf.Nodes); err != nil {
+				return errors.Wrapf(err, "check apphash failed after test command, %s", n.RunCmd)
+			}
+		}
+	}
+
+	return nil
+}
+
+func checkapphash(nodes map[string]*node.Node) error {
+	time.Sleep(time.Second * 1)
+	fmt.Printf("--> run all: %v \n", "checkapphash")
+	var apphash = make(map[string]struct{})
+	var lastBlockHeight int64
+	for _, v := range nodes {
+		u := fmt.Sprintf("%s/abci_info", v.RPCAddress)
+		resp, err := http.Get(u)
+		if err != nil {
+			return err
+		}
+		defer resp.Body.Close()
+		if resp.StatusCode != 200 {
+			respBytes, _ := ioutil.ReadAll(resp.Body)
+			return fmt.Errorf("post status not OK: %s, response body: %s", resp.Status, string(respBytes))
+		}
+		var info = struct {
+			JSONRPC string `json:"jsonrpc"`
+			ID      string `json:"id"`
+			Result  struct {
+				Response abciResponseInfo2 `json:"response"`
+			} `json:"result"`
+		}{}
+
+		err = json.NewDecoder(resp.Body).Decode(&info)
+		if err != nil {
+			return err
+		}
+		newLastBlockHeight, err := strconv.ParseInt(info.Result.Response.LastBlockHeight, 10, 64)
+		if err != nil {
+			return err
+		}
+		if lastBlockHeight == 0 {
+			lastBlockHeight = newLastBlockHeight
+		}
+		if lastBlockHeight == newLastBlockHeight {
```

Comment:
If some nodes are one block behind, app hash won't be included?

Reply:
I guess when the code was originally written this was not considered a problem. I'll try adding a wait-loop here?